src/renderer/dx/DxRenderer.cpp

```diff
@@ -997,6 +1000,49 @@ try
 
         _d2dRenderTarget->BeginDraw();
         _isPainting = true;
+
+        {
+            // Get the baseline for this font as that's where we draw from
+            DWRITE_LINE_SPACING spacing;
+            RETURN_IF_FAILED(_dwriteTextFormat->GetLineSpacing(&spacing.method, &spacing.height, &spacing.baseline));
+
+            // GH#5098: If we're rendering with cleartype text, we need to always
+            // render onto an opaque background. If our background's opacity is
+            // 1.0f, that's great, we can use that. Otherwise, we need to force the
+            // text renderer to render this text in grayscale. In
+            // UpdateDrawingBrushes, we'll set the backgroundColor's a channel to
+            // 1.0 if we're in cleartype mode and the background's opacity is 1.0.
+            // Otherwise, at this point, the _backgroundColor's alpha is <1.0.
+            //
+            // Currently, only text with the default background color uses an alpha
+            // of 0, every other background uses 1.0
+            //
+            // DANGER: Layers slow us down. Only do this in the specific case where
+            // someone has chosen the slower ClearType antialiasing (versus the faster
+            // grayscale antialiasing)
+            const bool usingCleartype = _antialiasingMode == D2D1_TEXT_ANTIALIAS_MODE_CLEARTYPE;
+            const bool usingTransparency = _defaultTextBackgroundOpacity != 1.0f;
+            // Another way of naming "bgIsDefault" is "bgHasTransparency"
+            const auto bgIsDefault = (_backgroundColor.a == _defaultBackgroundColor.a) &&
+                                     (_backgroundColor.r == _defaultBackgroundColor.r) &&
+                                     (_backgroundColor.g == _defaultBackgroundColor.g) &&
+                                     (_backgroundColor.b == _defaultBackgroundColor.b);
+            const bool forceGrayscaleAA = usingCleartype &&
+                                          usingTransparency &&
+                                          bgIsDefault;
+
+            // Assemble the drawing context information
+            _drawingContext = std::make_unique<DrawingContext>(_d2dRenderTarget.Get(),
+                                                               _d2dBrushForeground.Get(),

```

Comment:
I'm a little surprised that just setting the pointers to the brushes here in StartPaint works for the entire frame, even if the colors get changed. Presumably when the fg/bg brushes change, they're swapping in-place, so the pointer value doesn't change even when the color does?

Reply:
Correct. We honestly don't really need two brushes either. Probably only one.